### PR TITLE
[MPS] Migrate GLU to Metal

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #187833

The forward was initially built purely on the binary `TensorIterator`: split `self` into two halves and run an elementwise `a * sigmoid(b)`. This was dropped because it was ~2x slower than MPSGraph, as  for a last-dim split (the common SwiGLU/FFN layout) the two halves are strided views, so it routes through the generic strided binary
kernel.

For a contiguous source the tensor is collapsed around the split dim into `[outer, 2L]` (`L` = halved dim * inner dims) and a dense 2D grid is dispatched, reading the second half through a fixed `L`-element offset so there is no
per-element div/mod. The backward is a single fused pass that computes sigmoid
internally and writes both grad halves  The strided binary kernel is kept only as a fallback for non-contiguous inputs.

### Performance

Median us vs the MPSGraph baseline (Apple silicon; shapes from typical
SwiGLU/FFN sizes, measured with the script below):

| mode | geomean speedup | range |
| --- | --- | --- |
| forward | 1.03x | 0.98x - 1.13x |
| forward + backward | 1.21x | 1.11x - 1.28x |

<details>
<summary><code>bench_glu.py</code></summary>

```python
"""GLU forward / backward benchmark, mirroring bench_argminmax_pr.py format.

For each (shape, dtype, dim, mode) prints a JSON line with the median wall
time in microseconds. Pipe to a file to compare baseline (MPSGraph) vs
candidate (Metal). Median over 500 iterations with 20-iter warmup,
torch.mps.synchronize() bracketing each call.
"""

import json
import time

import torch
import torch.nn.functional as F

# Shapes pulled from typical SwiGLU/GLU FFN sizes plus a small square case so
# we measure both the launch-overhead-bound and the bandwidth-bound regimes.
# The doubled dim is the dim we'll split over.
CASES = [
    # (shape, dim) -- shape[dim] must be even
    ((1024, 1024), -1),
    ((4096, 4096), -1),
    ((4096, 4096), 0),
    # SwiGLU-shaped: (batch*seq, 2*hidden)
    ((2048, 8192), -1),
    ((4096, 11008), -1),  # llama-2 7B FFN intermediate
    # 4D activation
    ((8, 512, 64, 128), -1),
]

DTYPES = [torch.float32, torch.float16, torch.bfloat16]


def bench(fn, n=500, warmup=20):
    for _ in range(warmup):
        fn()
    s = []
    for _ in range(n):
        torch.mps.synchronize()
        t0 = time.perf_counter()
        fn()
        torch.mps.synchronize()
        s.append((time.perf_counter() - t0) * 1e6)
    return sorted(s)[n // 2]


def main():
    for shape, dim in CASES:
        for dt in DTYPES:
            x = torch.randn(shape, dtype=dt, device="mps")

            # Forward
            us_fwd = bench(lambda: F.glu(x, dim=dim))
            print(json.dumps({
                "op": "glu",
                "mode": "fwd",
                "shape": list(shape),
                "dim": dim,
                "dtype": str(dt).split(".")[1],
                "us": round(us_fwd, 1),
            }))

            # Forward + backward
            x_req = x.detach().requires_grad_(True)
            out_shape = list(shape)
            out_shape[dim] = shape[dim] // 2
            g = torch.randn(out_shape, dtype=dt, device="mps")

            def step():
                if x_req.grad is not None:
                    x_req.grad = None
                y = F.glu(x_req, dim=dim)
                y.backward(g)

            us_bwd = bench(step)
            print(json.dumps({
                "op": "glu",
                "mode": "fwd+bwd",
                "shape": list(shape),
                "dim": dim,
                "dtype": str(dt).split(".")[1],
                "us": round(us_bwd, 1),
            }))


if __name__ == "__main__":
    main()
```

</details>

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/ActivationKernel.metal [hunk 1]
```diff
@@ -324,6 +324,34 @@ REGISTER_BINARY_OP(sigmoid_backward, bfloat, bfloat);
 REGISTER_BINARY_OP(sigmoid_backward, float2, float2);
 REGISTER_BINARY_OP(sigmoid_backward, half2, half2);
 
+struct glu_functor {
+  template <typename T>
+  inline T operator()(const T a, const T b) {
+    const float bf = float(b);
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/ActivationKernel.metal:330]: Strict upcast? Interesting...

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -12057,17 +12056,15 @@
 - func: glu_backward.grad_input(Tensor grad_output, Tensor self, int dim, *, Tensor(a!) grad_input) -> Tensor(a!)
   python_module: nn
   dispatch:
-    CPU: glu_backward_cpu_out
+    CPU, MPS: glu_backward_cpu_out
```
Comment (Skylion007) [inline aten/src/ATen/native/native_functions.yaml]: Huh, why does only this go through CPU dispatch? Or I guess the stub is poorly named.

**Review by kurtamohler — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (malfet) [reply aten/src/ATen/native/native_functions.yaml]: Yeah, it's a bit weird, and somewhat wrong, as CPU first does sigmoid and then continues with the backward pass

**Comment by malfet** [comment]
Current approach indeed works, but it's slower than MPSGraph.. Need to think of few more optimizations here

**Comment by malfet** [comment]
@codex review this
